politeiad/cmd/legacypoliteia/convertcmd.go

```diff
@@ -0,0 +1,467 @@
+func execConvertCmd(args []string) error {
+	// Verify the git repo exists
+	if len(args) == 0 {
+		return fmt.Errorf("missing git repo argument")
+	}
+	gitRepo := util.CleanAndExpandPath(args[0])
+	if _, err := os.Stat(gitRepo); err != nil {
+		return fmt.Errorf("git repo not found: %v", gitRepo)
+	}
+
+	// Parse the CLI flags
+	err := convertFlags.Parse(args[1:])
+	if err != nil {
+		return err
+	}
+
+	// Clean the legacy directory path
+	*legacyDir = util.CleanAndExpandPath(*legacyDir)
+
+	// Verify the user ID
+	if *userID != "" {
+		_, err = uuid.Parse(*userID)
+		if err != nil {
+			return fmt.Errorf("invalid user id '%v': %v", *userID, err)
+		}
+	}
+
+	// Setup the legacy directory
+	err = os.MkdirAll(*legacyDir, filePermissions)
+	if err != nil {
+		return err
+	}
+
+	client, err := util.NewHTTPClient(false, "")
+	if err != nil {
+		return err
+	}
+
+	// Setup the cmd context
+	c := convertCmd{
+		client:       client,
+		gitRepo:      gitRepo,
+		legacyDir:    *legacyDir,
+		skipComments: *skipComments,
+		skipBallots:  *skipBallots,
+		ballotLimit:  *ballotLimit,
+		userID:       *userID,
+	}
+
+	// Convert the git proposals
+	return c.convertGitProposals()
+}
+
+// convertGitProposals converts the git proposals to tstore proposals, saving
+// the tstore proposals to disk as the conversion is finished.
+func (c *convertCmd) convertGitProposals() error {
+	// Build an inventory of all git proposal tokens
+	tokens, err := gitProposalTokens(c.gitRepo)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Found %v legacy git proposals\n", len(tokens))
+
+	// Parse git vote timestamps using git log. If parsed ballot is limited
+	// skip this.
+	var ts map[string]map[string]int64
+	if c.ballotLimit == 0 {
+		ts, err = parseVoteTimestamps(c.gitRepo)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Convert the data for each proposal into tstore supported types.
+	count := 1
+	for token := range tokens {
+		fmt.Printf("Converting proposal (%v/%v)\n", count, len(tokens))
+
+		// Get the path to the most recent version of the proposal.
+		// The version number is the directory name. We only import
+		// the most recent version of the proposal.
+		//
+		// Example path: [gitRepo]/[token]/[version]/
+		v, err := latestVersion(c.gitRepo, token)
+		if err != nil {
+			return err
+		}
+
+		version := strconv.FormatUint(v, 10)
+		proposalDir := filepath.Join(c.gitRepo, token, version)
+
+		// Convert git backend types to tstore backend types
+		recordMD, err := convertRecordMetadata(proposalDir)
+		if err != nil {
+			return err
+		}
+		files, err := convertFiles(proposalDir)
+		if err != nil {
+			return err
+		}
+		proposalMD, err := convertProposalMetadata(proposalDir)
+		if err != nil {
+			return err
+		}
+		voteMD, err := convertVoteMetadata(proposalDir)
+		if err != nil {
+			return err
+		}
+		userMD, err := convertUserMetadata(proposalDir)
+		if err != nil {
+			return err
+		}
+		// Populate user ID
```

Comment:
You turned this readable function into a tangled mess by adding all of the below logic inline.